### PR TITLE
New environment variable to change easily the publication mode

### PR DESCRIPTION
**Pull request by JLBuenoLopez-eProsima**
Several issues are related to the default publication mode set in `rmw_fastrtps` which is `ASYNCHRONOUS` publish mode.
This configuration can be changed only if the user defines a XML file and also sets the environment variable `RMW_FASTRTPS_USE_QOS_FROM_XML` to 1. This could be cumbersome to the user and also, as the issues show, prone to errors.

This pull request uses a new environment variable `RMW_FASTRTPS_PUBLICATION_MODE` so the user can change easily the publishing mode without needing to define a XML file.
This new environment variable can be used also in other RMW implementations and the values that have been considered are the following (this is open to discussion and more values can be given to this new environment variable):
* `ASYNCHRONOUS`: asynchronous publishing mode (default option in `rmw_fastrtps`).
* `SYNCHRONOUS`: synchronous publishing mode.
* `AUTO`: leave the rmw implementation to configure this parameter. In the case of `rmw_fastrtps` this value will take the value defined in the XML file or, failing this, the default value given in Fast DDS (`SYNCHRONOUS`).

### CODE REVIEW

**Comment by JLBuenoLopez-eProsima** [comment]
@hidmic,

If you agree that it is interesting to use this environment variable in other rmw implementations, I assume that I should document this new variable in some other more general repository. I would be grateful if you could tell me where.

File: README.md
Comment (hidmic) [inline README.md:49]: @JLBuenoLopez-eProsima I'm somewhat inclined to stick to `RMW` for a prefix: 
```suggestion
Another way to change easily the publication mode is to use the environment variable `RMW_PUBLICATION_MODE`.
```

File: rmw_fastrtps_shared_cpp/src/participant.cpp
Comment (hidmic) [inline rmw_fastrtps_shared_cpp/src/participant.cpp:203]: @JLBuenoLopez-eProsima consider using `RMW_SET_ERROR` instead of logging.

File: rmw_fastrtps_shared_cpp/src/participant.cpp
Comment (hidmic) [inline rmw_fastrtps_shared_cpp/src/participant.cpp:212]: @JLBuenoLopez-eProsima if it's neither, a warning log stating that the given value is unknown and the default will take over would be nice.

**Review by hidmic — COMMENTED** [review]
@JLBuenoLopez-eProsima to configure modes like this would be an interesting feature, but to put it in the `rmw` API spec we have to unambiguously define what each mode entails, decide on what the best granularity for configuration would be, and how should implementations that lack support for some (or all) of these modes behave.

Either a design document or a draft update to [`rmw`](https://github.com/ros2/rmw) documentation could kickstart the discussion.

**Comment by hidmic** [comment]
CC @wjwwood @ivanpauno for thoughts on making (or not making) this an `rmw` feature.

**Comment by ivanpauno** [comment]
> ROS_PUBLICATION_MODE

Considering this option is only supported in `rmw_fastrtps`, I would name the variable `RMW_FASTRTPS_PUBLICATION_MODE`.
Otherwise, this sounds good to me.

**Comment by wjwwood** [comment]
> Considering this option is only supported in rmw_fastrtps, I would name the variable RMW_FASTRTPS_PUBLICATION_MODE.
Otherwise, this sounds good to me.

Agreed. I do not think we want `rmw` or ROS specific features flagged on and off with env variables... This variable will affect all publishers, and will change the behavior significantly of all publish calls. It should be used very carefully imo. I'm only in favor of it because it affects on `rmw_fastrtps_cpp` and because it is already possible by setting two other environment variables.

**Comment by EduPonz** [comment]
Our line of thought was following the one discussed in ros2/ros2#255, #343, and ros2/rmw_cyclonedds#89. From those discussions, it can be concluded that asynchronous publishing is indeed a requirement for the middlewares, and that having a synchronous mode is optional. In fact, #343 was not merged because of this, even though synchronous publication has better latency performance that can compete with that of other vendors.

IMHO there are cases when applications benefit more from one than the other. We have in fact dealt with issues regarding this over and over, where we recommend changing the publication mode. However, this is currently cumbersome since it involves XML configuration files + environment variables. In that sense, I think it's better to make it as easy as possible for the user to make the change (because why not). In fact, I think this could be part of the upper-level API, so it can be configured programatically from the  beginning. In that sense, maybe the publish mode configuration could be publisher based, publication based, context based, etc.

Sure right now it's only Fast DDS who can make the switch, but that is just because no other vendors (apart from connext) meet the requirement regarding this. However, that might change in the future, so having an mechanism already in place would benefit everyone. Furthermore, user have easier access to upper-layer documentations, so it'll simply get more visibility upstream than just being a Fast DDS specific thing (when it should not). This will also entail letting the user know when the variable does not apply when using other RMWs, which is probably a good thing, since it let's the user know about the middleware behaviour, instead of being different from one RMW to the other without the user knowing or even wanting.

Also, there are other cases were environment variables change ROS behaviour. An example of it would be `ROS_LOCALHOST_ONLY`, so it won't be a first. I agree with @wjwwood that it would require careful usage, but that does not mean we should limit users' options. Instead, a good documentation will help them make the best decision for their use-case, instead of us making a decision for them that maybe they don't want, and that is different from RMW to RMW.

File: rmw_fastrtps_shared_cpp/src/participant.cpp
Comment (JLBuenoLopez-eProsima) [reply rmw_fastrtps_shared_cpp/src/participant.cpp:217]: Done

File: rmw_fastrtps_shared_cpp/src/participant.cpp
Comment (JLBuenoLopez-eProsima) [reply rmw_fastrtps_shared_cpp/src/participant.cpp]: Done

File: README.md
Comment (JLBuenoLopez-eProsima) [reply README.md]: For the moment, I have changed the name to the one proposed by @ivanpauno. If later other rmw implementations start to use a similar variable to configure the publication mode we can consider updating this code.

**Comment by wjwwood** [comment]
> Sure right now it's only Fast DDS who can make the switch, but that is just because no other vendors (apart from connext) meet the requirement regarding this. However, that might change in the future, so having an mechanism already in place would benefit everyone. 

Yes, but in my opinion that should be a QoS setting that you set when you create a publisher, not an environment variable.

> An example of it would be ROS_LOCALHOST_ONLY, so it won't be a first.

For me the important difference (which makes `ROS_LOCALHOST_ONLY` ok and something like `RMW_ASYNCHRONOUS_PUBLISHING` not ok imo), is that the `ROS_LOCALHOST_ONLY` doesn't (significantly or concretely) affect the behavior of the API calls, whereas this change can significantly affect the blocking behavior of `DataWriter::write()`.

**Comment by EduPonz** [comment]
> Yes, but in my opinion that should be a QoS setting that you set when you create a publisher, not an environment variable.

I completely agree with this vision. It'd be great to be able to create a publisher with either publication mode. The thing is that currently this is not possible, that's why we wanted some way to make it easier for the user to do so. I think this is related to some discussions held here and there, about how to design an API that is agnostic of the specific middleware, while at the same time lets the user really tune the middleware (https://discourse.ros.org/t/restricting-communication-between-robots/2931/15?u=eduponz).

I'm very much ok with setting the name to something like `RMW_FASTRTPS_PUBLICATION_MODE` if such APIs will exist in the future in a more general and ordered manner (maybe in Galactic). In fact, I think it'd be really great for users and vendors to have some set of requirements in terms of features, performance, configuration paramentes, etc. I think the way the Quality Levels were worked out is a good approach to these kinds of problems, specially because the whenever anyone would make a RMW switch, everything that that entails would be well documented and quite visible.

Such set of requirements may be worked out in the middleware working group, and could also help the decision of the default middleware. They'd most importantly be very beneficial for the community as a whole, since they would push development towards the real end-users needs. The requirement would also help designing what QoS are configured programatically; if a feature is required, it should be configurable via QoS. IMO this would be a very interesting discussion to have if you guys are up to moving it somewhere else. Doing so, we would avoid the somewhat unorthodox way of configuring with XML we have right now.

> For me the important difference (which makes ROS_LOCALHOST_ONLY ok and something like RMW_ASYNCHRONOUS_PUBLISHING not ok imo), is that the ROS_LOCALHOST_ONLY doesn't (significantly or concretely) affect the behavior of the API calls, whereas this change can significantly affect the blocking behavior of DataWriter::write().

You are completely right, the change will significantly affect the behaviour of the call. However, I do not think that's a bad thing in itself, as long as the implications are well documented. @JLBuenoLopez-eProsima , what do you think about explaining a bit more the difference between the modes in the README file?

File: rmw_fastrtps_shared_cpp/src/participant.cpp
Comment (fujitatomoya) [inline rmw_fastrtps_shared_cpp/src/participant.cpp:213]:
```suggestion
      } else if (strcmp(env_value, "ASYNCHRONOUS") != 0) {
```

I think empty case `""` is not admissible based on the design.

Comment (JLBuenoLopez-eProsima) [reply rmw_fastrtps_shared_cpp/src/participant.cpp:213]: The environment variable is read using `rcutils_get_env()`:

```
const char *
rcutils_get_env(const char * env_name, const char ** env_value)
{
  RCUTILS_CAN_RETURN_WITH_ERROR_OF("some string error");

  if (NULL == env_name) {
    return "argument env_name is null";
  }
  if (NULL == env_value) {
    return "argument env_value is null";
  }

  // TODO(Suyash458): getenv is deprecated on Windows; consider using getenv_s instead
  *env_value = getenv(env_name);

  if (NULL == *env_value) {
    *env_value = "";
  }
  return NULL;
}
```
It uses function `getenv` that returns `NULL` if the variable does not exist. Below, if `NULL` has been returned, `rcutils_get_env()` returns empty string `""`. Therefore, to avoid the warning log appearance when the environment variable is not defined, this condition should also be checked.

**Comment by JLBuenoLopez-eProsima** [comment]
@EduPonz,

> @JLBuenoLopez-eProsima , what do you think about explaining a bit more the difference between the modes in the README file?

Please read last commit.

File: rmw_fastrtps_shared_cpp/src/participant.cpp
Comment (fujitatomoya) [reply rmw_fastrtps_shared_cpp/src/participant.cpp:213]: that makes sense, appreciate for the explanation.

File: README.md
Comment (EduPonz) [inline README.md:53]:
```suggestion
Setting this mode implies that when the publisher invoques the write operation, the data is copied into a queue, a notification about the addition to the queue is performed, and control of the thread is returned to the user before the data is actually sent.
A background thread (asynchronous thread) is in turn in charge of consuming the queue and sending the data to every matched reader.
```

File: README.md
Comment (EduPonz) [inline README.md:55]:
```suggestion
Setting this mode implies that the data is sent directly within the context of the user thread.
This entails that any blocking call occurring during the write operation would block the user thread, thus preventing the application with continuing its operation.
It is important to note that this mode typically yields higher throughput rates at lower latencies, since the notification and context switching between threads is not present.
```

File: README.md
Comment (EduPonz) [inline README.md:58]:
```suggestion
By default, `rmw_fastrtps` assumes this variable to be `ASYNCHRONOUS`, which is the publication mode in use if this environment variable is not set.
```

**Review by EduPonz — CHANGES_REQUESTED** [review]
Some suggestions in the docs

**Review by hidmic — COMMENTED** [review]
@JLBuenoLopez-eProsima how do `RMW_FASTRTPS_USE_QOS_FROM_XML` and  `RMW_FASTRTPS_PUBLICATION_MODE` interact? 

I'll kickstart CI in the meantime:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/12928/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/7880/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/10647/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/12878/)

**Comment by JLBuenoLopez-eProsima** [comment]
@hidmic,

I have been studying the test that is failing in MacOS (`TestNodeConstructionDestruction__rmw_[implementation].create_with_internal_errors`). I have launched these tests locally in MacOS using directly the Rolling Ridley versions and I have seen that the test is flaky. Also, not only is failing in `rmw_fastrtps_cpp` and `rmw_fastrtps_dynamic_cpp` implementations, but also in `rmw_cyclonedds_cpp`. The test fails with an uncaught exception of type `std::__1::system_error: mutex lock failed: Invalid  argument`.

Therefore, the MacOS failures seem unrelated to this pull request and a more in-depth analysis of this test should be done.

**Comment by hidmic** [comment]
> RMW_FASTRTPS_USE_QOS_FROM_XML has priority over RMW_FASTRTPS_PUBLICATION_MODE.

Great. Thanks for the clarification. Would you mind adding that sentence to the documentation?

**Comment by hidmic** [comment]
> Therefore, the MacOS failures seem unrelated to this pull request and a more in-depth analysis of this test should be done.

Indeed. I can see flakes in MacOS nightlies.

**Review by hidmic — APPROVED** [review]
LGTM once https://github.com/ros2/rmw_fastrtps/pull/470#issuecomment-721945928 is addressed.

**Comment by JLBuenoLopez-eProsima** [comment]
> LGTM once [#470 (comment)](https://github.com/ros2/rmw_fastrtps/pull/470#issuecomment-721945928) is addressed.

Last commit explains the precedence between both environment variables.

It also clarifies the relationship between the XML file and `RMW_FASTRTPS_USE_QOS_FROM_XML` taking into account [#466 (comment)](https://github.com/ros2/rmw_fastrtps/pull/466/files#r517322626).

**Comment by EduPonz** [comment]
Hi @hidmic ,

Do you think this is good to go then? IMHO, it provides the users of `rmw_fastrtps` with an easy a quick solution for a recurring problem that otherwise is more tedious to solve, while at the same time documenting the capabilities regarding the publication mode better. I think it can only do good to have it.

**Review by hidmic — APPROVED** [review]
Thanks @JLBuenoLopez-eProsima. @EduPonz yes, I'm onboard with this change. @wjwwood has a good point though. Hopefully, at some point, the `rmw` API is flexible enough to support such configurations and environment variables are no longer needed. Proposals are most welcome.

**Comment by hidmic** [comment]
Another approval would be nice @wjwwood @ivanpauno @mjeronimo.

File: README.md
Comment (mjeronimo) [inline README.md:62]: Do you really mean "assumes" here? That is, is there an implicit assumption that the value is ASYNCHRONOUS? Or, does this mean that rmw_fastrtps sets the value, by default, to ASYNCHRONOUS? I suppose the code will clarify, but the description should be clear.

**Review by mjeronimo — APPROVED** [review]

File: README.md
Comment (wjwwood) [reply README.md:62]: I think it must be "if `RMW_FASTRTPS_PUBLICATION_MODE` is not set, then `rmw_fastrtps_cpp` behaves as if it were set to `ASYNCHRONOUS`.

**Review by wjwwood — APPROVED** [review]

File: README.md
Comment (JLBuenoLopez-eProsima) [reply README.md]: Done in 2bb6f5fbb1e7f971ab941c944998d0ca20a23c88

**Comment by EduPonz** [comment]
Hi everyone,

The suggestion made by @wjwwood was accounted for in 2bb6f5f. Is there anything else to be done before merging?

**Comment by hidmic** [comment]
Nothing else to be added.

**Comment by EduPonz** [comment]
Thanks so much to everybody to push this further!

**Pull request merged**
